composer/utils/collect_env.py

```diff
@@ -0,0 +1,245 @@
+    from torch.cuda import get_device_name as accel_device_name
+    from torch.cuda import is_available as cuda_available
+    TORCH_AVAILABLE = True
+except (ImportError, NameError, AttributeError, OSError):
+    TORCH_AVAILABLE = False
+
+# Check if Composer is installed
+try:
+    import composer
+    from composer.utils import dist
+    COMPOSER_AVAILABLE = True
+except (ImportError, NameError, AttributeError, OSError):
+    COMPOSER_AVAILABLE = False
+
+COMPOSER_OPEN_ISSUE_URL = "https://github.com/mosaicml/composer/issues/new/choose"
+EXCEPTION_MSG = f"If you believe this exception was raised due to a Composer bug, " + \
+                f"file a bug report at: {COMPOSER_OPEN_ISSUE_URL}\n" + \
+                f"Please include the following environment report:\n"
+
+
+# Same convention as Torch collect_env, create a namedtuple to track collected fields
+class ComposerEnv(NamedTuple):
+    composer_version: str
+    number_of_nodes: int
+    host_processor_model_name: str
+    host_processor_core_count: int
+    accelerators_per_node: int
+    accelerator_model_name: str
+
+
+# Helper functions to get Composer environment information
+def get_composer_version() -> str:
+    """Query the Composer version."""
+    return str(composer.__version__)
+
+
+def get_num_nodes() -> int:
+    """Query the number of nodes."""
+    return int(dist.get_world_size() / dist.get_local_world_size())
+
+
+def get_host_processor_name() -> str:
+    """Query the host processor name."""
+    cpu_info = cpuinfo.get_cpu_info()
+    return str(cpu_info.get('brand_raw', "CPU"))
+
+
+def get_host_processor_cores() -> int:
+    """Determine the number of physical host processor cores."""
+    return psutil.cpu_count(logical=False)
+
+
+def get_accel_per_node() -> int:
+    """Determine the number of accelerators per node."""
+    return dist.get_local_world_size() if cuda_available() else 0
+
+
+def get_accel_model_name() -> str:
+    """Query the accelerator name."""
+    return accel_device_name(None) if cuda_available() else "N/A"
+
+
+# Excepthook wrapper, wraps default excepthook and prints env info
+def _excepthook_wrapper(type, value, tb) -> None:
+    sys.__excepthook__(type, value, tb)
+    print("\n-------------------\n" + EXCEPTION_MSG)
+    print_env()
+
+
+# Public function to register excethook wrapper
+def configure_excepthook() -> None:
+    """Collect and print system information when :func:`sys.excepthook` is called.
+
+    To override the default :func:`sys.excepthook` with the custom except hook:
+
+    .. testsetup::
+
+        import sys
+        from composer.utils.collect_env import configure_excepthook
+
+    .. doctest::
+
+        >>> configure_excepthook()
+        >>> sys.excepthook 
+        <function _excepthook_wrapper at ...>
+    """
+    sys.excepthook = _excepthook_wrapper
+
+
+# Public function to restore original excepthook
+def restore_excepthook() -> None:
+    """Restore default :func:`sys.excepthook` behavior.
```

Comment:
Instead of restoring the default excepthook, could we restore the excepthook that we patched over?